cmd/goauth/main.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2022 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package main
+
+import (
+	"context"
+	"encoding/base64"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"time"
+
+	"github.com/GoogleCloudPlatform/artifact-registry-go-tools/pkg/auth"
+)
+
+const help = `
+Handle Go authentication with Google Cloud Artifact Registry Go Repositories.
+
+Add to your GOAUTH environment variable:
+
+  export GOAUTH="sh -c 'GOPROXY=direct go run github.com/GoogleCloudPlatform/artifact-registry-go-tools/cmd/goauth@latest <location>'"
+
+To support multiple locations, add the command multiple times to the GOAUTH variable (semicolon-separated).
+
+For more details, see https://pkg.go.dev/cmd/go@master#hdr-GOAUTH_environment_variable`
+
+func main() {
+	jsonKey := flag.String("json_key", "", "path to the json key of the service account used for this location. Leave empty to use the oauth token instead.")
+	hostPattern := flag.String("host_pattern", "%s-go.pkg.dev", "Artifact Registry server host pattern, where %s will be replaced by a location string.")
+
+	flag.Parse()
+
+	location := flag.Arg(0)
+	if location == "" {
+		fmt.Println(help)
+		return
+	}
+
+	err := handleLocation(location, *jsonKey, *hostPattern)
+	if err != nil {
+		log.Println(err)
+		os.Exit(1)
+	}
+}
+
+func handleLocation(location string, jsonKeyPath string, hostPattern string) error {
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	host := fmt.Sprintf(hostPattern, location)
```

Comment:
I may be being a bit thick 

but if 
`GOPROXY=https://europe-west2-go.pkg.dev`. 

Location would be `https://europe-west2-go.pkg.dev` for example this would generate a host of `https://europe-west2-go.pkg.dev-go.pkg.dev` which is invalid.

This should also be testing if the location is ok to be processed by this auth

Reply:
Not sure why GOPROXY environment variable would be related to the location 🤔 . The location is provided explicitly and is not derived from the GOPROXY variable, from what I see.

From my POV, the location is the region name at the Google Cloud.

The implementation follows the existing logic: https://github.com/smartpricer/artifact-registry-go-tools/blob/189aa77d79ad83ed0f0f458a38cd5b8a73c91686/pkg/netrc/netrc.go#L112

If there is an issue, I assume both location would need to be changed.

Reply:
Okay, I understand you are refering to:

> Before the first HTTPS fetch, the go command will invoke each GOAUTH command in the list with no additional arguments and no input. *If the server responds with any 4xx code, the go command will invoke the GOAUTH commands again with the URL as an additional command-line argument and the HTTP Response to the program's stdin.* If the server responds with an error again, the fetch fails: a URL-specific GOAUTH will only be attempted once per fetch.

But the current version is basically ignoring the second case, as it assumes the first one is working. 

I see it is a reasonable extension for security, to prevent sending the authentication to non-Google proxies.

Reply:
I added a safety check, but from my POV this is nothing that could happen in real life, because this would require for the location argument to be missing, which would then already fail the first call.

Reply:
Yeah i misunderstood the location thing as in my version of this arg(0) is the arg go provide on the second invocation
```
If the server responds with any 4xx code, the go command will invoke the
GOAUTH commands again with the URL as an additional command-line argument
and the HTTP Response to the program's stdin.
```

Hence my validation reasoning was behind this.